tagstudio/src/core/library.py

```diff
@@ -2315,3 +2175,340 @@ def sort_fields(self, entry_id: int, order: list[int]) -> None:
         entry.fields = sorted(
             entry.fields, key=lambda x: order.index(self.get_field_attr(x, "id"))
         )
+
+
+class SpecialFlag:
+    only_untagged: bool
+    only_no_author: bool
+    only_empty: bool
+    only_missing: bool
+
+    def __init__(self, query: str):
+        self.only_no_author: bool = "no author" in query or "no artist" in query
+        self.only_untagged: bool = "untagged" in query or "no tags" in query
+        self.only_empty: bool = "empty" in query or "no fields" in query
+        self.only_missing: bool = "missing" in query or "no file" in query
+
+
+class Filter:
+    def __init__(self, lib: Library) -> None:
+        # self.lib = lib
+        self.ext_list = lib.ext_list
+        self.entries = lib.entries
+        self.is_exclude_list = lib.is_exclude_list
+        self.get_field_attr = lib.get_field_attr
+        self._tag_strings_to_id_map = lib._tag_strings_to_id_map
+        self.get_tag_cluster = lib.get_tag_cluster
+        self.library_dir = lib.library_dir
+        self.get_field_obj = lib.get_field_obj
+        self.missing_files = lib.missing_files
+        self.default_fields = lib.default_fields
+
+        def generate_field_id_to_name_map(self) -> dict[int, str]:
+            map: dict = {}
+            for field in self.default_fields:
+                map[field.get("id")] = field.get("name").lower()
+            return map
+
+        self._field_id_to_name_map: dict = generate_field_id_to_name_map(self)
+
+    def filter_results(
+        self, query: str, split_query: list[dict], search_mode: SearchMode
+    ) -> list[tuple[ItemType, int]]:
+        # start_time = time.time()
+        query = query.strip().lower()
```

Comment:
other places are using `.casefold()` instead of `.lower()`, it would be good to unify the approach since both of these methods are behaving differently in some edge cases (eg. German's scharfes S).